ci/check-protobuf-stability.sh

```diff
@@ -35,42 +45,52 @@ function check_non_lf_protos() {
   readonly BUF_IMAGE_TMPDIR="$(mktemp -d)"
   trap 'rm -rf ${BUF_IMAGE_TMPDIR}' EXIT
 
-  # The `SYSTEM_PULLREQUEST_TARGETBRANCH` environment variable is defined by
-  # Azure Pipelines; in order to run this script locally, define it beforehand
-  # as the branch being targeted. For example:
-  #
-  #   SYSTEM_PULLREQUEST_TARGETBRANCH=main bash -x ci/check-protobuf-stability.sh
-  #
-  TARGET="${SYSTEM_PULLREQUEST_TARGETBRANCH:-main}"
-  echo "The target branch is '${TARGET}'."
+  echo "Checking protobufs against tag '${BUF_TAG_TO_CHECK}'"
+  for buf_module in "${BUF_MODULES_AGAINST_STABLE[@]}"; do
+    (eval "$(dev-env/bin/dade assist)" ; buf breaking --config "${buf_module}" --against "$BUF_TAG_TO_CHECK")
+  done
 
-  # For `main` and PRs targeting `main`, we simply check against the most recent
-  # stable tag.
-  #
-  # For PRs targeting release branches, we should really check against
-  # all the most recent stable tags reachable from either the current branch or
-  # from previous release branches (say, against both `1.17.1` and `1.16.2`
-  # created after the `release/1.17.x` branch).
-  # Instead, we approximate by checking only against the most recent stable tag
-  # reachable from the current branch, under the assumption that if a lesser
-  # release branch contains a protobuf change, then it will also be present in
-  # higher ones either through a shared commit or a back-port from `main`.
+}
+
+case "${1:---stable}" in
+-h | --help)
+  cat <<USAGE
+Usage: ./check-protobuf-stability.sh [options]
+
+Options:
+  -h, --help: shows this help
+  --stable:     check against the latest stable tag (default)
+  --head:       check against the head commit
+USAGE
+  exit
+  ;;
+--stable)
+  echo "Using stable target"
+  # Check against the most recent stable tag.
   #
-  # Finally, this check does not need to run on release branch commits because
+  # This check does not need to run on release branch commits because
   # they are built sequentially, so no conflicts are possible and the per-PR
   # check is enough.
+  readonly RELEASE_BRANCH_REGEX="^release/.*"
   GIT_TAG_SCOPE=""
   if [[ "${TARGET}" =~ ${RELEASE_BRANCH_REGEX} ]]; then
     GIT_TAG_SCOPE="--merged"
   fi
-
   readonly LATEST_STABLE_TAG="$(git tag ${GIT_TAG_SCOPE} | grep -v "snapshot" | sort -V | tail -1)"
-  echo "Checking protobufs against tag '${LATEST_STABLE_TAG}'"
-  for buf_module in "${BUF_MODULES_AGAINST_STABLE[@]}"; do
-    (eval "$(dev-env/bin/dade assist)" ; buf breaking --config "${buf_module}" --against ".git#tag=${LATEST_STABLE_TAG}")
-  done
-
-}
-
+  BUF_TAG_TO_CHECK=".git#tag=${LATEST_STABLE_TAG}"
+  ;;
+--head)
+  # Check against the head of the target branch.
+  #
+  # This check ensures that backwards compatibility is never broken on the target branch,
+  # which is stricter than guaranteeing compatibility between release tags.
```

Comment:
My understanding, from my discussions with @fabiotudone-da, was that we specifically wanted to be able to break compatibility in-between releases (which I think is a bad idea, but not my call). With this PR, that would no longer be possible. If we're checking that PRs never break compatibility with their parent, and we assume not-breaking-backwards-compatibility is a transitive property (which the comment here strongly implies), then what do we still gain by ever comparing to tags at all?

If we're willing to have this stricter policy, then we can simplify this a bit and just always check against all parents, on both PR builds and main builds. (This will do what you want because Azure builds merge commits, so the target branch is one of the parents.)

You can list the parents of the current commit with
```
git log --pretty=%P -1
```

Reply:
> My understanding, from my discussions with @fabiotudone-da, was that we specifically wanted to be able to break compatibility in-between releases (which I think is a bad idea, but not my call). With this PR, that would no longer be possible. If we're checking that PRs never break compatibility with their parent, and we assume not-breaking-backwards-compatibility is a transitive property (which the comment here strongly implies), then what do we still gain by ever comparing to tags at all?

An upcoming change will introduce a check override mechanism via commit message trailers.

Reply:
The idea is to add an exception to this check, working on that as soon as I'm done with this story (KVL-1111).
So we would have a way of going around this check.
AFAIK there is no way of skipping the stable check, so while this check is stricter and it would basically invalidate the stable check, we still want to run the stable check to close the loophole where we explicitly skip this check.